File: backend/src/MessageFromServer.kt
package `in`.vilik

data class KafkaTopicPartition(
  val id: Int,
  val replicas: List<Int>,
  val inSyncReplicas: List<Int>,
  val leader: Int
)

data class KafkaTopic(
  val id: String,
  val isInternal: Boolean,
  val partitions: List<KafkaTopicPartition>
)

data class PartitionOffsets(
  val topic: String,
  val partition: Int,
  val earliest: Long,
  val latest: Long
)

data class ConsumerGroupOffsetForPartition(
  val topic: String,
  val partition: Int,
  val offset: Long
)

data class KafkaConsumerGroup(
  val id: String,
  val offsets: List<ConsumerGroupOffsetForPartition>
)

data class KafkaRecord(
  val topic: String,
  val partition: Int,
  val offset: Long,
  val timestamp: Long,
  val key: String?,
  val value: String
)

sealed class MessageFromServer(
  val type: Type
) {
  enum class Type {
    STATUS_CONNECTED,
    STATUS_CONNECTING,
    STATUS_DISCONNECTED,
    REFRESH_TOPICS,
    REFRESH_CONSUMER_GROUPS,
    REFRESH_TOPIC_OFFSETS,
    RECEIVE_RECORDS,
    SUBSCRIBED_TO_RECORDS_OF_TOPIC,
    UNSUBSCRIBED_FROM_RECORDS_OF_TOPIC,
    SUBSCRIBED_TO_OFFSETS_OF_CONSUMER_GROUP,
    UNSUBSCRIBED_FROM_OFFSETS_OF_CONSUMER_GROUP,
    SUBSCRIBED_TO_OFFSETS_OF_TOPIC,
    UNSUBSCRIBED_FROM_OFFSETS_OF_TOPIC
  }
}

class StatusConnected(
  val environmentId: String
) : MessageFromServer(Type.STATUS_CONNECTED)

class StatusConnecting(
  val environmentId: String
) : MessageFromServer(Type.STATUS_CONNECTING)

class StatusDisconnected(
  val reason: String
) : MessageFromServer(Type.STATUS_DISCONNECTED)

class RefreshTopics(
  val topics: List<KafkaTopic>
) : MessageFromServer(Type.REFRESH_TOPICS)

class RefreshConsumerGroups(
  val consumerGroups: List<KafkaConsumerGroup>
) : MessageFromServer(Type.REFRESH_CONSUMER_GROUPS)

class RefreshTopicOffsets(
  val offsets: List<PartitionOffsets>
) : MessageFromServer(Type.REFRESH_TOPIC_OFFSETS)

class SubscribedToRecordsOfTopic(
  val topic: String
) : MessageFromServer(Type.SUBSCRIBED_TO_RECORDS_OF_TOPIC)

class UnsubscribedFromRecordsOfTopic(
  val topic: String
) : MessageFromServer(Type.UNSUBSCRIBED_FROM_RECORDS_OF_TOPIC)

class ReceiveRecords(
  val records: List<KafkaRecord>
) : MessageFromServer(Type.RECEIVE_RECORDS)

class SubscribedToOffsetsOfConsumerGroup(
  val groupId: String
) : MessageFromServer(Type.SUBSCRIBED_TO_OFFSETS_OF_CONSUMER_GROUP)

class UnsubscribedFromOffsetsOfConsumerGroup(
  val groupId: String
) : MessageFromServer(Type.UNSUBSCRIBED_FROM_OFFSETS_OF_CONSUMER_GROUP)

class SubscribedToOffsetsOfTopic(
  val topic: String
) : MessageFromServer(Type.SUBSCRIBED_TO_OFFSETS_OF_TOPIC)

class UnsubscribedFromOffsetsOfTopic(
  val topic: String
) : MessageFromServer(Type.UNSUBSCRIBED_FROM_OFFSETS_OF_TOPIC)

File: src/kafka/message-from-server.ts
export enum MessageFromServerType {
  STATUS_CONNECTED = "STATUS_CONNECTED",
  STATUS_CONNECTING = "STATUS_CONNECTING",
  STATUS_FAILED_TO_CONNECT = "STATUS_FAILED_TO_CONNECT",
  STATUS_DISCONNECTED = "STATUS_DISCONNECTED",
  REFRESH_TOPICS = "REFRESH_TOPICS",
  REFRESH_CONSUMER_GROUPS = "REFRESH_CONSUMER_GROUPS",
  REFRESH_TOPIC_OFFSETS = "REFRESH_TOPIC_OFFSETS",
  RECEIVE_RECORDS = "RECEIVE_RECORDS",
  SUBSCRIBED_TO_RECORDS_OF_TOPIC = "SUBSCRIBED_TO_RECORDS_OF_TOPIC",
  UNSUBSCRIBED_FROM_RECORDS_OF_TOPIC = "UNSUBSCRIBED_FROM_RECORDS_OF_TOPIC",
}

export enum DisconnectReason {
  CLIENT_REQUESTED_DISCONNECT = "CLIENT_REQUESTED_DISCONNECT",
  ALL_WEBSOCKET_SESSIONS_CLOSED = "ALL_WEBSOCKET_SESSIONS_CLOSED",
  APPLICATION_EXITING = "APPLICATION_EXITING",
}

export interface KafkaTopicPartition {
  id: number;
  replicas: number[];
  inSyncReplicas: number[];
  leader: number;
}

export interface KafkaTopic {
  id: string;
  isInternal: boolean;
  partitions: KafkaTopicPartition[];
}

export interface PartitionOffsets {
  topic: string;
  partition: number;
  earliest: number;
  latest: number;
}

export interface ConsumerGroupOffsetForPartition {
  topic: string;
  partition: number;
  offset: number;
}

export interface KafkaConsumerGroup {
  id: string;
  offsets: ConsumerGroupOffsetForPartition[];
}

export interface KafkaRecord {
  topic: string;
  partition: number;
  offset: number;
  timestamp: number;
  key: string | null;
  value: string;
}

export interface StatusConnected {
  type: MessageFromServerType.STATUS_CONNECTED;
  environmentId: string;
}

export interface StatusConnecting {
  type: MessageFromServerType.STATUS_CONNECTING;
  environmentId: string;
}

export interface StatusFailedToConnect {
  type: MessageFromServerType.STATUS_FAILED_TO_CONNECT;
  environmentId: string;
  error: string;
}

export interface StatusDisconnected {
  type: MessageFromServerType.STATUS_DISCONNECTED;
  environmentId: string;
  reason: DisconnectReason;
}

export interface RefreshTopics {
  type: MessageFromServerType.REFRESH_TOPICS;
  topics: KafkaTopic[];
}

export interface RefreshConsumerGroups {
  type: MessageFromServerType.REFRESH_CONSUMER_GROUPS;
  consumerGroups: KafkaConsumerGroup[];
}

export interface RefreshTopicOffsets {
  type: MessageFromServerType.REFRESH_TOPIC_OFFSETS;
  offsets: PartitionOffsets[];
}

export interface SubscribedToRecordsOfTopic {
  type: MessageFromServerType.SUBSCRIBED_TO_RECORDS_OF_TOPIC;
  topic: string;
}

export interface UnsubscribedFromRecordsOfTopic {
  type: MessageFromServerType.UNSUBSCRIBED_FROM_RECORDS_OF_TOPIC;
  topic: string;
}

export interface ReceiveRecords {
  type: MessageFromServerType.RECEIVE_RECORDS;
  records: KafkaRecord[];
}

export type MessageFromServer =
  | StatusConnected
  | StatusConnecting
  | StatusFailedToConnect
  | StatusDisconnected
  | RefreshTopics
  | RefreshConsumerGroups
  | RefreshTopicOffsets
  | SubscribedToRecordsOfTopic
  | UnsubscribedFromRecordsOfTopic
  | ReceiveRecords;
